pkg/controller/volume/persistentvolume/index.go

```diff
@@ -124,13 +131,18 @@ func (pvIndex *persistentVolumeOrderedIndex) findByClaim(claim *v1.PersistentVol
 				if volumeQty.Cmp(requestedQty) < 0 {
 					continue
 				}
+				// filter out mismatching volumeModes
```

Comment:
You could check this at the beginning of the for loop since it's checked in both scenarios.